### PR TITLE
[JENKINS-57852] - Deprecate Java 7 targets support in Plugin POM

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Follow-up to https://groups.google.com/forum/#!topic/jenkinsci-dev/SdMOMKs7XIQ
Should unblock #200 and other similar PRs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,15 +36,15 @@ In order to use the new POM:
...
+    <version>3.43</version> <!-- or whatever the newest version available is -->
     <relativePath />
   </parent>
 ```
 * Override the needed properties, e.g.:
 ```xml
   <properties>
-    <jenkins.version>1.609.1</jenkins.version>
-    <java.level>7</java.level>
+    <jenkins.version>2.64.1</jenkins.version>
```
Comment (jglick) [inline README.md]:
```suggestion
    <jenkins.version>2.60.1</jenkins.version>
```

File: src/it/sample-plugin/pom.xml [hunk 2]
```diff
@@ -13,8 +13,8 @@
     <version>1.0-SNAPSHOT</version>
     <packaging>hpi</packaging>
     <properties>
-        <jenkins.version>2.32.3</jenkins.version>
-        <java.level>7</java.level>
+        <jenkins.version>2.64.3</jenkins.version>
```
Comment (jglick) [inline src/it/sample-plugin/pom.xml]:
```suggestion
        <jenkins.version>2.60.3</jenkins.version>
```

**Comment by oleg-nenashev** [comment]
thanks @jglick , quickhack 🤦‍♂

File: README.md [hunk 3]
```diff
@@ -53,6 +54,32 @@ If you had a `jar:test-jar` execution, delete it and add to `properties`:
 <no-test-jar>false</no-test-jar>
 ```
 
+## Java support
+
+The plugin POM is designed for plugin builds with JDK 8 or above,
+but target `java.level` for a plugin may differ from a JDK version used for the build.
+Starting from Plugin POM `3.44`, support of Java 7 targets in Plugin POM is deprecated,
+`java.level=8` and `jenkins.version>2.59` are expected to be used for the most of the plugins.
```
Comment (jglick) [inline README.md]:
```suggestion
`java.level=8` and `jenkins.version>2.59` are expected to be used for most plugins.
```

File: README.md [hunk 4]
```diff
@@ -53,6 +54,32 @@ If you had a `jar:test-jar` execution, delete it and add to `properties`:
...
+
+### Using deprecated Java targets
+
+If recent plugin POM versions are required for a plugin with older baselines,
+a developer can use dependency version system properties to downgrade components incompatible  with Java 7.
```
Comment (jglick) [inline README.md]:
```suggestion
a developer can use dependency version system properties to downgrade components incompatible with Java 7.
```

File: pom.xml [hunk 5]
```diff
@@ -872,6 +872,36 @@
         <java.level>8</java.level>
       </properties>
     </profile>
+    <profile>
+      <id>deprecated-java.level-7</id>
+      <activation>
+        <property>
+          <name>java.level</name>
+          <value>7</value>
```
Comment (jglick) [inline pom.xml]: There may be some plugins using 6 out there still, though I suppose they are not using a modern POM either…

File: src/it/not-newest-java-level/pom.xml [hunk 6]
```diff
@@ -14,7 +14,7 @@
     <packaging>hpi</packaging>
     <properties>
         <jenkins.version>2.89</jenkins.version>
-        <java.level>7</java.level>
+        <java.level>8</java.level>
```
Comment (jglick) [inline src/it/not-newest-java-level/pom.xml]: Contradicts the intention of the IT added in #91. If there is no usable example, better to just delete this IT.

File: pom.xml (hunk 5 above)
Comment (oleg-nenashev) [reply pom.xml]: There is no good way to do conditional profiles or to chain them. Hit the same in Java 11 support PRs. So I just did the check for Java 7, but can copy-paste if needed

Comment (jglick) [reply pom.xml]: I think it is good enough, just noting this.

File: pom.xml [hunk 7]
```diff
@@ -872,6 +872,36 @@
...
+            <executions>
+              <execution>
+                <phase>initialize</phase>
+                <id>warn-deprecated-java.level</id>
+                <goals>
+                  <goal>run</goal>
+                </goals>
+                <configuration>
+                  <tasks>
+                    <echo>WARNING: Plugin POM defines java.level=${java.level} which is deprecated in the current Plugin POM version. See https://github.com/jenkinsci/plugin-pom/blob/master/README.md#java-support</echo>
```
Comment (jglick) [inline pom.xml]: As an aside, maybe for #133, we ought to be deprecating this property and introducing a new one which expects an actual Java specification version—thus `1.8`, `11`, etc.

Comment (oleg-nenashev) [reply pom.xml]: Agreed

Comment (casz) [reply pom.xml]: I agree this would properly be a good time to do it 😓

Comment (jglick) [reply pom.xml]: Maybe possible to combine the tasks? Have a profile issuing a warning with two different clauses if you have `java.level` defined at all, and introduce the new property defaulting to `1.${java.level}`. Just a thought; it is not a blocker for this PR.

**Comment by oleg-nenashev** [comment]
Will merge it taking the feedback in the mailing list

**Pull request merged**